# Introduction to MATLAB

## Section 1: MATLAB Basics

Let's begin by typing stuff into the console and seeing what happens.

### Section 1.1: Arithmetic

Just like with every other programming language, you can use MATLAB as a calculator.  Type the commands you see in the cells marked `In` into your console and press `Enter`; your output should match the `Out` cell directly beneath it.

In [1]:
1 + 3


ans =

     4



In [2]:
2 * 4


ans =

     8



In [3]:
1/2


ans =

    0.5000



As you would expect, MATLAB respects the order of operations you are used to from math:

In [4]:
20 + 2 / 1 * 2


ans =

    24



In [7]:
(20 + 2) / 1 * 2


ans =

    44



You can save values to variables and do calculations with them.  Note the `;` at the end of each of the lines below: adding a `;` to a statement suppresses output to the console.

In [ ]:
x = 20;
y = 7;
x * y - 13

Note the output: MATLAB stores the result of your last entered command in the variable `ans`.  

In [8]:
ans


ans =

    44



#### Exercise 1.1

You can pull up the documentation on a function by typing `doc <function>` in the console.  Look up the documentation for the `mod` function and answer the following questions.

1. How can you tell when a number is odd?
2. How can you tell when a number is divisible by 10?


### Section 1.2 : Matrices

MATLAB is all about matrices.  Everything is a fucking matrix.  There are several ways to make matrices.  For example, if we wanted to make a row vector of the first 10 integers, we can do...

In [5]:
a = [1, 2, 3, 4, 5, 6, 7, 8, 9]


a =

     1     2     3     4     5     6     7     8     9



The `size` function returns the dimensions of an object.  For example:

In [6]:
size(a)


ans =

     1     9



This means that `a` is a matrix of 1 row and 9 columns.  

You can access elements of a matrix by passing its index:

In [26]:
a(5)


ans =

     5



You can use the `end` keyword to access the last element of a matrix:

In [27]:
a(end)


ans =

     9



You can also access a range of values:

In [28]:
a(2:5)


ans =

     2     3     4     5



You can create a column vector by separating the row entires with a `;`:

In [30]:
b = [10; 11; 12; 13; 14; 15; 16; 17; 18;]


b =

    10
    11
    12
    13
    14
    15
    16
    17
    18



When we call `size` on `b`:

In [31]:
size(b)


ans =

     9     1



When the desired matrix is evenly spaced, we can take advantage of the `:` operator.  The syntax to use the `:` operator is `start:end`.  This will return a row vector of values from `start` to `end`.  For example, to create matrix `a` using the `:` operator, we can do:

In [12]:
a = 1:9


a =

     1     2     3     4     5     6     7     8     9



We can use the transpose (as in matrix transpose) operator `'` to create a column vector.  For example, we can create matrix `b` with the following command:

In [35]:
b = (10:18)'


b =

    10
    11
    12
    13
    14
    15
    16
    17
    18



The `:` operator also accepts an argument for the step size.  The syntax is `start:step:end`.  By default, `step` is set to 1.  Using the `:` operator, we can list the even numbers between 1 and 10 like so:

In [24]:
0:2:10


ans =

     0     2     4     6     8    10



The step size doesn't have to be an integer:

In [19]:
1:0.1:2


ans =

    1.0000
    1.1000
    1.2000
    1.3000
    1.4000
    1.5000
    1.6000
    1.7000
    1.8000
    1.9000
    2.0000



And we can combine the output of the `:` operator with the transpose operator `'`:

In [20]:
(13:5:39)'


ans =

    13
    18
    23
    28
    33
    38



It may be the case that you know how many values you'd like but you don't know how to evenly space them.  In this case, you can use the `linspace` function.  The syntax for `linspace` is `linspace(start, end, n)`; `linspace` will return a **lin**early **space**d vector of `n` elements from `start` to `end`.  For example, if we wanted a list of 5 elements between 13 and 15:

In [21]:
linspace(13, 15, 5)


ans =

   13.0000   13.5000   14.0000   14.5000   15.0000



You can create matrices of higher dimensions by combining `;` and `,`:

In [22]:
c = [1:3; 4:6; 7:9]


c =

     1     2     3
     4     5     6
     7     8     9



We can access elements of matrix by supply a row and column number.  For instnace, to access the 8 in `c`, we can use:

In [38]:
c(3, 2)


ans =

     8



In [ ]:
To access

We can access all the elements along a given dimension by passing `:` to the index.  For instance, to access all the elements in the 2nd column, we can do:

In [39]:
c(:, 2)


ans =

     2
     5
     8



Or the 3rd row:

In [40]:
c(3, :)


ans =

     7     8     9



We can also reshape matrices using the `reshape` function.  The syntax for `reshape` is `reshape(mat, [dim_1, dim_2, ...]`, where `mat` is the matrix to be resized and `[dim_1, dim_2, ... ]` is the row vector of dimensions to resize `mat` to.  For example, we can turn `a` and `b` into 3 by 3 matrices:

In [36]:
a = reshape(a, [3, 3])


a =

     1     4     7
     2     5     8
     3     6     9



In [37]:
b = reshape(b, [3, 3])


b =

    10    13    16
    11    14    17
    12    15    18



#### Exercise 1.2

1. Use the `:` operator to count backwards from 10.  Save it to the variable 
2. Create a column vector `d` that contains the first 64 integers.  Reshape it into a 3-dimensional square matrix.
3. What happens when you call `c(end)`?  Why?

### Section 1.3: More About Matrices

MATLAB has a few matrix creation methods